Checking the existence of an item

Is there a more elegant way to check if an item exists, given a name?

Currently, items['ANonExistentItem'] will cause an exception. I'd have to rescue the exception to proceed.

It would be good I think if items['ANonExistentItem'] will simply return nil.

Sorry, I've just tested `items.include? 'Something'`. It doesn't matter whether `Something` is an existing item or non existent, it still raised an error:

```
Switch                TestItem1
```

```ruby
rule 'rule1 ' do
  on_start
  run do
    logger.info("on? #{items['TestItem1'].on?}")
    logger.info("include? #{items.include?('TestItem1')}")
  end
end
```

```
07:04:02.816 [INFO ] [.OpenHAB::Core::DSL::Rule::RuleConfig] - on? false
07:04:02.846 [ERROR] [ript.internal.ScriptEngineManagerImpl] - Error during evaluation of script 'file:/openhab/conf/automation/jsr223/ruby/personal/test.rb': org.jruby.embed.EvalFailedException: java.lang.NumberFormatException: Character T is neither a decimal digit number, decimal point, nor "e" notation exponential mark.
```

--- a/docs/usage/items.md
+++ b/docs/usage/items.md
@@ -15,6 +15,7 @@ All items can be accessed as an enumerable the `items` method.
 | Method             | Description                                                                    |
 | ------------------ | ------------------------------------------------------------------------------ |
 | []                 | Get a specific item by name, this syntax can be used to dynamically load items |
+| include?           | Check to see if an item with the given name exists                             |
 | enumerable methods | All methods [here](https://ruby-doc.org/core-2.5.0/Enumerable.html)            |
 
 ## Examples
@@ -29,6 +30,8 @@ Switch SwitchTest "Test Switch"
 ```ruby
 logger.info("Item Count: #{items.count}")  # Item Count: 2
 logger.info("Items: #{items.sort_by(&:label).map(&:label).join(', ')}")  #Items: Test Dimmer, Test Switch' 
+logger.info("DimmerTest exists? #{items.include? 'DimmerTest'}") # DimmerTest exists? true
+logger.info("StringTest exists? #{items.include? 'StringTest'}") # StringTest exists? false
 ```
 
 ```ruby

--- a/docs/usage/items.md
+++ b/docs/usage/items.md
@@ -41,6 +41,15 @@ rule 'Use dynamic item lookup to increase related dimmer brightness when switch 
 end
 ```
 
+```ruby
+rule 'search for a suitable item' do
+  on_start
+  triggered do
+    (items['DimmerTest'] || items['SwitchTest'])&.on # Send ON to DimmerTest if it exists, otherwise send it to SwitchTest
+  end
+end
+```
+
 ## All Items
 Item types have methods added to them to make it flow naturally within the a ruby context.  All methods of the OpenHAB item are available plus the additional methods described below.
 

--- a/features/items.feature
+++ b/features/items.feature
@@ -162,3 +162,13 @@ Feature:  Rule languages supports groups
       """
     When I deploy the rules file
     Then It should log "SwitchTest Received Update" within 5 seconds
+
+  Scenario: Check for item existence
+    Given code in a rules file
+      """
+      logger.info("DimmerTest include? #{items.include? 'DimmerTest'}")
+      logger.info("SimmerTest include? #{items.include? 'SimmerTest'}")
+      """
+    When I deploy the rules file
+    Then It should log 'DimmerTest include? true' within 5 seconds
+    And It should log 'SimmerTest include? false' within 5 seconds

--- a/features/items.feature
+++ b/features/items.feature
@@ -168,7 +168,9 @@ Feature:  Rule languages supports groups
       """
       logger.info("DimmerTest include? #{items.include? 'DimmerTest'}")
       logger.info("SimmerTest include? #{items.include? 'SimmerTest'}")
+      logger.info("SimmerTest item nil? #{items['SimmerTest'].nil?}")
       """
     When I deploy the rules file
     Then It should log 'DimmerTest include? true' within 5 seconds
     And It should log 'SimmerTest include? false' within 5 seconds
+    And It should log 'SimmerTest item nil? true' within 5 seconds

--- a/lib/openhab/core/dsl/items/items.rb
+++ b/lib/openhab/core/dsl/items/items.rb
@@ -22,6 +22,8 @@ module OpenHAB
             item = $ir.getItem(name)
             # rubocop: enable Style/GlobalVars
             (item.is_a? GroupItem) ? nil : item
+          rescue Java::OrgOpenhabCoreItems::ItemNotFoundException
+            nil
           end
 
           # Returns true if the given item name exists

--- a/lib/openhab/core/dsl/items/items.rb
+++ b/lib/openhab/core/dsl/items/items.rb
@@ -23,6 +23,14 @@ module OpenHAB
             # rubocop: enable Style/GlobalVars
             (item.is_a? GroupItem) ? nil : item
           end
+
+          # Returns true if the given item name exists
+          # @param name [String] Item name to check
+          # @return [Boolean] true if the item exists, false otherwise
+          def include?(name)
+            !$ir.getItems(name).empty?
+          end
+          alias key? include?
         end
 
         java_import org.openhab.core.items.GroupItem

